### PR TITLE
Add llvm-amdgpu compiler to C++ config

### PR DESCRIPTION
**Pull request by cgmb**
This change adds the llvm-amdgpu compiler for building C++. The raison d'etre of llvm-amdgpu is really the HIP language (which has partial support right now under the CUDA category), but I'm pretty new to Compiler Explorer and the combination of adding a compiler and expanding support for a new language at the same time sounded a little intimidating. I figured I'd just start by adding the compiler to C++. While the primary purpose of llvm-amdgpu is for HIP language code, it's still fairly common for it to be used for vanilla C and C++.

As I understand it, the names that I'm entering into in the change to compiler-explorer here depend on what I've named them in PRs to other repos. I don't think I can fully test this change until the changes in those other repos are merged, so I'm just opening this one as a draft.

Related:
https://github.com/compiler-explorer/compiler-workflows/pull/9
https://github.com/compiler-explorer/clang-builder/pull/40
https://github.com/compiler-explorer/infra/pull/809

### CODE REVIEW

**Comment by mattgodbolt** [comment]
> As I understand it, the names that I'm entering into in the change to compiler-explorer here depend on what I've named them in PRs to other repos. 

Partly true: the names don't matter but:
- the paths need to agree
- once set, the IDs need to never change, so choose carefully.

File: etc/config/c++.amazon.properties [hunk 1]
```diff
@@ -1,4 +1,4 @@
-compilers=&gcc86:&icc:&icx:&clang:&clangx86trunk:&rvclang:&wasmclang:&cl:&cross:&ellcc:&zapcc:&djggp:www.godbolt.ms@443:&armclang32:&armclang64:&zigcxx:&cxx6502
+compilers=&gcc86:&icc:&icx:&clang:&clangx86trunk:&llvm-amdgpu:&rvclang:&wasmclang:&cl:&cross:&ellcc:&zapcc:&djggp:www.godbolt.ms@443:&armclang32:&armclang64:&zigcxx:&cxx6502
```
Comment (mattgodbolt) [inline etc/config/c++.amazon.properties:1]: In other places we call this by the compiler (llvm is the library we build and link as a client library elsewhere). I already made this rename in other places, can you do it here please?

File: etc/config/c++.amazon.properties [hunk 2]
```diff
@@ -392,6 +392,36 @@ compiler.clang_reflection.semver=(reflection)
 compiler.clang_reflection.options=-std=c++20 -freflection-ts -stdlib=libc++
 compiler.clang_reflection.notification=Experimental Reflection Support; see <a href="https://github.com/cplusplus/reflection-ts" target="_blank" rel="noopener noreferrer">github<sup><small class="fas fa-external-link-alt opens-new-window" title="Opens in a new window"></small></sup></a>
 
+group.llvm-amdgpu.compilers=llvm-amdgpu-trunk:llvm-amdgpu-40502:llvm-amdgpu-50002:llvm-amdgpu-50103:llvm-amdgpu-50203
```
Comment (mattgodbolt) [inline etc/config/c++.amazon.properties:395]: Ditto here: does `clang-amdgpu-` work? that is more descriptive as an ID but the actual project we're building is `clang-rocm-*`:

```
Sep/21 14:46 admin-node~/infra (main|✔) $ ls -ld /opt/compiler-explorer/clang-rocm-*
drwxr-xr-x 8 root root 6144 Sep 21 14:30 /opt/compiler-explorer/clang-rocm-4.5.2
drwxr-xr-x 8 root root 6144 Sep 21 14:34 /opt/compiler-explorer/clang-rocm-5.0.2
drwxr-xr-x 8 root root 6144 Sep 21 14:38 /opt/compiler-explorer/clang-rocm-5.1.3
drwxr-xr-x 8 root root 6144 Sep 21 14:42 /opt/compiler-explorer/clang-rocm-5.2.3
lrwxrwxrwx 1 root root   25 Sep 21 14:19 /opt/compiler-explorer/clang-rocm-trunk -> clang-rocm-trunk-20220921
drwxr-xr-x 8 root root 6144 Sep 21 14:17 /opt/compiler-explorer/clang-rocm-trunk-20220921
```

are the paths.

File: etc/config/c++.amazon.properties [hunk 3]
```diff
@@ -392,6 +392,36 @@ compiler.clang_reflection.semver=(reflection)
...
+group.llvm-amdgpu.groupName=llvm-amdgpu x86-64
+group.llvm-amdgpu.instructionSet=amd64
+group.llvm-amdgpu.baseName=x86-64 clang
+group.llvm-amdgpu.isSemVer=true
+group.llvm-amdgpu.compilerType=clang
+group.llvm-amdgpu.unwiseOptions=-march=native
+group.llvm-amdgpu.supportsPVS-Studio=true
+
+compiler.llvm-amdgpu-trunk.exe=/opt/compiler-explorer/llvm-amdgpu-trunk/bin/clang++
+compiler.llvm-amdgpu-trunk.semver=(amd-stg-open)
```
Comment (mattgodbolt) [inline etc/config/c++.amazon.properties:407]: I do'nt feel too strongly about this, but calling it `turnk` might help others who don't know what `amd-stg-open` is understand it's a nightly bulit version of the latest version of the code.

**Review by mattgodbolt — CHANGES_REQUESTED** [review]
Some name and path changes; but the structure looks sound! thanks. We should be able to get this live later today :)

File: etc/config/c++.amazon.properties (hunk 3 above)
Comment (cgmb) [reply etc/config/c++.amazon.properties:407]: That's true, although I think giving users the keyword `amd-stg-open` makes it easier for them to find more information about the build. It's also the terminology that the AMD devs are familiar with, which might make it easier to report compiler bugs.

File: etc/config/c++.amazon.properties (hunk 2 above)
Comment (cgmb) [reply etc/config/c++.amazon.properties:395]: `clang-rocm` is probably fine. The compiler is installed via the `rocm-llvm` package, and it describes itself as:
```
AMD clang version 14.0.0 (https://github.com/RadeonOpenCompute/llvm-project roc-5.2.3 22324 d6c88e5a78066d5d7a1e8db6c5e3e9884c6ad10e)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/rocm-5.2.3/llvm/bin
```

File: etc/config/c++.amazon.properties (hunk 1 above)
Comment (cgmb) [reply etc/config/c++.amazon.properties:1]: Sure.

File: etc/config/c++.amazon.properties (hunk 3 above)
Comment (RubenRBS) [reply etc/config/c++.amazon.properties]: THen it's all good. There is going to be so ordering issues with the current semvers of all compilers, but that's something we can fix once this is merged, I'll add myself a reminder to do so :)

File: etc/config/c++.amazon.properties [hunk 4]
```diff
@@ -392,6 +392,36 @@ compiler.clang_reflection.semver=(reflection)
...
+group.clang-rocm.compilers=clang-rocm-trunk:clang-rocm-40502:clang-rocm-50002:clang-rocm-50103:clang-rocm-50203
+group.clang-rocm.intelAsm=-mllvm --x86-asm-syntax=intel
+group.clang-rocm.options=--gcc-toolchain=/opt/compiler-explorer/gcc-7.2.0
+group.clang-rocm.groupName=ROCm clang x86-64
+group.clang-rocm.instructionSet=amd64
+group.clang-rocm.baseName=x86-64 clang
+group.clang-rocm.isSemVer=true
+group.clang-rocm.compilerType=clang
+group.clang-rocm.unwiseOptions=-march=native
+group.clang-rocm.supportsPVS-Studio=true
```
Comment (RubenRBS) [inline etc/config/c++.amazon.properties:404]: As of a few days ago, we're asking compiler groups to provide license info
`group.clang-rocm.licensePreamble` for custom text to be shown,
`group.clang-rocm.licenseName` for the license name,
`group.clang-rocm.licenseLink` for either a link to the license or other relevant URL.

Can you provide those? All three are optional if you don't have anything for the preamble, for example

Comment (cgmb) [reply etc/config/c++.amazon.properties:404]: It's a mix of LLVM Apache 2 and NCSA. There's two links:

https://github.com/RadeonOpenCompute/llvm-project/blob/amd-stg-open/LICENSE.TXT
https://github.com/RadeonOpenCompute/ROCm-Device-Libs/blob/amd-stg-open/LICENSE.TXT

Comment (cgmb) [reply etc/config/c++.amazon.properties:406]: I've added the second link in a comment.

**Comment by RubenRBS** [comment]
This looks good, merging now & it should be live soon-ish. Thanks for your patience :)

**Pull request merged**